### PR TITLE
Prioritize serverSide between browserSide validation

### PR DESCRIPTION
**Pull request by roshan-gh**
This pull request addresses the issue of deduplicating serverSide and browserSide validation errors with prioritizing them. 

**Changes Made:**

- Disabling warnings when serverSide validation is active
- Disabling json schema validation when serverSide validation is active
- Displaying serverSide validation errors after browserSide validation errors have been fixed

This change will improve the user experience by reducing the number of error messages that are displayed

### CODE REVIEW

**Comment by bripkens** [comment]
Some usage feedback:

image

The schema source is now shown twice in this dialog.

image

The same is true for all kind of suggestions in the editor. And actually, we shouldn't show these `schema.json` based suggestions when server-side validation is active. Because the chosen distribution may not support them.

image

The errors/warnings state is now consistent. The editor is showing warnings, while the error console does not.

image

Similar situation to the one shown above.

File: packages/otelbin/src/components/monaco-editor/otelCollectorConfigValidation.ts
Comment (bripkens) [inline packages/otelbin/src/components/monaco-editor/otelCollectorConfigValidation.ts:175]: **Suggestion:** This is a really uncommon code pattern outside of JSX. Please consider using regular `if` statements outside of JSX.

File: packages/otelbin/src/contexts/EditorContext.tsx
Comment (bripkens) [inline packages/otelbin/src/contexts/EditorContext.tsx:88]: **Question:** Why do we need a new context for this? It should be possible to always identify this from either the `useUrlState` result or `useServerSideValidation`. Contexts make it harder to maintain the code.

File: packages/otelbin/src/components/monaco-editor/parseYaml.ts
Comment (bripkens) [inline packages/otelbin/src/components/monaco-editor/parseYaml.ts:112]: **Question:** What is `path` being used for?

File: packages/otelbin/src/contexts/EditorContext.tsx
Comment (roshan-gh) [reply packages/otelbin/src/contexts/EditorContext.tsx:88]: @bripkens I made it as context to avoid code duplication for using useUrlState.

File: packages/otelbin/src/components/monaco-editor/otelCollectorConfigValidation.ts
Comment (roshan-gh) [reply packages/otelbin/src/components/monaco-editor/otelCollectorConfigValidation.ts:175]: @bripkens sure, thank you

File: packages/otelbin/src/components/monaco-editor/parseYaml.ts
Comment (roshan-gh) [reply packages/otelbin/src/components/monaco-editor/parseYaml.ts:112]: @bripkens I added it for the initial proposal (separating serverSide and browserSide) with comparing error path. and didn't remove it, I thought it will be useful for the future features.

Comment (bripkens) [reply packages/otelbin/src/components/monaco-editor/parseYaml.ts:112]: Let's not keep dead code around. If we need it down the road, we can always restore it from PRs we haven't merged/code we have changed :)

File: packages/otelbin/src/contexts/EditorContext.tsx
Comment (bripkens) [reply packages/otelbin/src/contexts/EditorContext.tsx:88]: If this is only about code duplication, then there are also approaches with less of a runtime impact.

```ts
export function useServerSideValidationEnabled(): Boolean {
  const [{ distro, distroVersion }] = useUrlState([distroBinding, distroVersionBinding]);
  return !!distro && !!distroVersion;
}
```

IMO context makes it harder to understand the code, because it requires more jumping around the code/understanding of the exact React tree existing at runtime.

**Comment by roshan-gh** [comment]
@bripkens Addressed the comments and pushed necessary changes.

**Comment by bripkens** [comment]
One last thing we should resolve before the merge is that the warning state is inconsistent. If we keep the warnings due to the technical issues we haven't overcome yet, let's at least be consistent, i.e., also present them at the bottom of the screen.

image

**Comment by mmanciop** [comment]
Converting to draft as we have not figured out yet how to reset the validation model for Monaco, and without that quirk tackled, we would not ship it.

**Comment by roshan-gh** [comment]
@bripkens @mmanciop Great news! We finally have it done, and it works exactly as pull request description.

**Review by bripkens — COMMENTED** [review]
Generally the changes are looking fine. Nice to see how the updated library is making this a whole lot easier! ❤️ 

The only thing missing here really are tests. Doesn't feel right that we don't have a single test case for this. Can you please add some @roshan-gh?

**Comment by roshan-gh** [comment]
@bripkens tests for custom validation with including useServerSideValidationEnabled hook and also test for useServerSideValidationEnabled hook itself pushed and commited.

**Review by bripkens — APPROVED** [review]
LGTM 👍 

Great to have this sorted out ❤️ !

**Pull request merged**
